crates/sui-verifier/src/id_leak_verifier.rs

```diff
@@ -154,52 +208,32 @@ impl<'a> TransferFunctions for IDLeakAnalysis<'a> {
 
 impl<'a> AbstractInterpreter for IDLeakAnalysis<'a> {}
 
-/// Certain Sui Framework functions can safely take a `ID` by value
-fn is_call_safe_to_leak(verifier: &IDLeakAnalysis, function_handle: &FunctionHandle) -> bool {
-    let m = verifier
-        .binary_view
-        .module_handle_at(function_handle.module);
-    let is_framework =
-        verifier.binary_view.address_identifier_at(m.address) == &SUI_FRAMEWORK_ADDRESS;
-    if !is_framework {
-        return false;
-    }
-
-    // sui::object::delete
-    (verifier.binary_view.identifier_at(m.name) == OBJECT_MODULE_NAME
-        && verifier
-            .binary_view
-            .identifier_at(function_handle.name)
-            .as_str()
-            == "delete") ||
-    // sui::transfer::delete_child_object
-    (verifier.binary_view.identifier_at(m.name).as_str() == "transfer"
-            && verifier
-                .binary_view
-                .identifier_at(function_handle.name)
-                .as_str()
-                == "delete_child_object")
-}
-
 fn call(
     verifier: &mut IDLeakAnalysis,
     function_handle: &FunctionHandle,
 ) -> Result<(), ExecutionError> {
-    let guaranteed_safe = is_call_safe_to_leak(verifier, function_handle);
     let parameters = verifier
         .binary_view
         .signature_at(function_handle.parameters);
-    for _ in 0..parameters.len() {
-        if verifier.stack.pop().unwrap() == AbstractValue::ID && !guaranteed_safe {
-            return Err(verification_failure(
-                "ID leaked through function call.".to_string(),
-            ));
-        }
-    }
+    verifier.stack_popn(parameters.len());
 
     let return_ = verifier.binary_view.signature_at(function_handle.return_);
-    for _ in 0..return_.0.len() {
-        verifier.stack.push(AbstractValue::NonID);
+    let function = verifier.resolve_function(function_handle);
+    if FRESH_ID_FUNCTIONS
+        .iter()
+        .any(|makes_fresh| function == *makes_fresh)
+    {
+        if return_.0.len() != 1 {
+            debug_assert!(false, "{:?} should have a single return value", function);
+            return Err(ExecutionError::from_kind(
+                ExecutionFailureStatus::InvariantViolation,
+            ));
+        }
+        verifier.stack.push(AbstractValue::Fresh);
+    } else {
+        verifier
+            .stack
+            .extend(vec![AbstractValue::Other; return_.0.len()]);
```

Comment:
Is it worth considering a `stock_pushn` that's implemented using `resize`?

```rust
fn stack_pushn(&mut self, n: usize, val: AbstractValue) {
    let new_len = self.stack.len() + n;
    self.stack.resize(new_len, val);
}
```

Reply:
ah, that's neat. Sure I'll do that, thanks!